refactor(dao) : LetterSound #1677

SnehaHS65 opened the pull request:
closes #1677

The version of webapp in pom.xml is still 2.4.11 and the db migration script in this name already exists. If you could change the version, I can create db script.

jo-elimu commented:
> The version of webapp in pom.xml is still 2.4.11 and the db migration script in this name already exists. If you could change the version, I can create db script.

@Sneha65 Done: https://github.com/elimu-ai/webapp/releases

jo-elimu commented on src/main/java/ai/elimu/web/content/word/WordCsvExportController.java:
```diff
@@ -57,23 +57,23 @@ public void handleRequest(
-                letterSoundCorrespondenceJsonObject.put("id", letterSound.getId());
+            for (LetterSound letterSound : word.getLetterSounds()) {
+                JSONObject letterSoundJsonObject = new JSONObject();
+                letterSoundJsonObject.put("id", letterSound.getId());
                 String[] lettersArray = new String[letterSound.getLetters().size()];
                 for (int i = 0; i < lettersArray.length; i++) {
                     lettersArray[i] = letterSound.getLetters().get(i).getText();
                 }
-                letterSoundCorrespondenceJsonObject.put("letters", lettersArray);
+                letterSoundJsonObject.put("letters", lettersArray);
                 String[] soundsArray = new String[letterSound.getSounds().size()];
                 for (int i = 0; i < soundsArray.length; i++) {
                     soundsArray[i] = letterSound.getSounds().get(i).getValueIpa();
                 }
-                letterSoundCorrespondenceJsonObject.put("sounds", soundsArray);
-                letterSoundCorrespondenceJsonObject.put("usageCount", letterSound.getUsageCount());
-                letterSoundCorrespondencesJsonArray.put(index, letterSoundCorrespondenceJsonObject);
+                letterSoundJsonObject.put("sounds", soundsArray);
+                letterSoundJsonObject.put("usageCount", letterSound.getUsageCount());
+                letterSoundsJsonArray.put(index, letterSoundJsonObject);
```
@coderabbitai Can you show an example of what that would look like?

SnehaHS65 commented on src/main/java/ai/elimu/web/content/peer_review/LetterSoundPeerReviewEventCreateController.java:
@jo-elimu , I don't know if I should change this because this is a URL, please let me know

jo-elimu replied:
@SnehaHS65 Thanks for pointing that out, because it looks like the original URL was wrong. I'll suggest a correction below.

SnehaHS65 commented:
@jo-elimu , can you please review and approve my PR

jo-elimu commented on src/main/java/ai/elimu/tasks/LetterSoundUsageCountScheduler.java:
```diff
@@ -0,0 +1,69 @@
+        List<Word> words = wordDao.readAll();
+        logger.info("words.size(): " + words.size());
+        for (Word word : words) {
+            logger.info("word.getText(): " + word.getText());
```
@SnehaHS65 This file is not in sync with the most recent changes on the `main` branch: https://github.com/elimu-ai/webapp/pull/1732/files#diff-82052f8d582b9a0720ed15bc313d8c8fa17f12e2cf7ec8677aa3ebe828e22a24

jo-elimu replied:
@SnehaHS65 Reminder to not change from `logger.debug` to `logger.info`.

jo-elimu commented on src/main/java/ai/elimu/tasks/LetterSoundUsageCountScheduler.java:
```diff
@@ -0,0 +1,69 @@
+
+        // Update the values previously stored in the database
+        for (LetterSound letterSound : letterSoundDao.readAll()) {
+            logger.info("letterSound.getId(): " + letterSound.getId());
+            logger.info("letterSound Letters: \"" + letterSound.getLetters().stream().map(Letter::getText).collect(Collectors.joining()) + "\"");
+            logger.info("letterSound Sounds: /" + letterSound.getSounds().stream().map(Sound::getValueIpa).collect(Collectors.joining()) + "/");
+            logger.info("letterSound.getUsageCount() (before update): " + letterSound.getUsageCount());
+            
+            int newUsageCount = 0;
+            if (letterSoundFrequencyMap.containsKey(letterSound.getId())) {
+                newUsageCount = letterSoundFrequencyMap.get(letterSound.getId());
+            }
+            logger.info("newUsageCount: " + newUsageCount);
```
@SnehaHS65 Please update this code to include the recent changes at https://github.com/elimu-ai/webapp/pull/1732/files#diff-82052f8d582b9a0720ed15bc313d8c8fa17f12e2cf7ec8677aa3ebe828e22a24

jo-elimu commented on src/main/java/ai/elimu/web/content/MainContentController.java:
```diff
@@ -119,7 +119,7 @@ public String handleRequest(
         model.addAttribute("letterCount", letterDao.readCount());
         model.addAttribute("soundCount", soundDao.readCount());
-        model.addAttribute("letterSoundCorrespondenceCount", letterSoundDao.readCount());
+        model.addAttribute("letterSoundCount", letterSoundDao.readCount());
```
This change also needs to be reflected in the controller's frontend file at `src/main/webapp/WEB-INF/jsp/content/main.jsp`.

<kbd>![](https://github.com/user-attachments/assets/613c0c76-21d2-4b61-960d-2622a0b679b5)</kbd>

jo-elimu commented on src/main/java/ai/elimu/web/content/peer_review/LetterSoundPeerReviewEventCreateController.java:
```diff
@@ -63,9 +63,9 @@ public String handleSubmit(
         
         if (!EnvironmentContextLoaderListener.PROPERTIES.isEmpty()) {
-            String contentUrl = "https://" + EnvironmentContextLoaderListener.PROPERTIES.getProperty("content.language").toLowerCase() + ".elimu.ai/content/letterSoundCorrespondence/edit/" + letterSoundContributionEvent.getLetterSound().getId();
+            String contentUrl = "https://" + EnvironmentContextLoaderListener.PROPERTIES.getProperty("content.language").toLowerCase() + ".elimu.ai/content/letterSound/edit/" + letterSoundContributionEvent.getLetterSound().getId();
```
```suggestion
            String contentUrl = "https://" + EnvironmentContextLoaderListener.PROPERTIES.getProperty("content.language").toLowerCase() + ".elimu.ai/content/letter-sound/edit/" + letterSoundContributionEvent.getLetterSound().getId();
```

jo-elimu commented on src/main/java/ai/elimu/web/content/sound/SoundEditController.java:
```diff
@@ -52,7 +52,7 @@ public String handleRequest(Model model, @PathVariable Long id) {
         model.addAttribute("soundContributionEvents", soundContributionEventDao.readAll(sound));
         
-        model.addAttribute("letterSoundCorrespondences", letterSoundDao.readAll());
+        model.addAttribute("letterSounds", letterSoundDao.readAll());
```
@Sneha65 At http://localhost:8080/webapp/content/sound/edit/1, I cannot see the list of letter-sounds anymore:

<kbd>![](https://github.com/user-attachments/assets/1c71fe13-aafb-49f1-bb99-eca1e8e96756)</kbd>

So you would probably have to update the attribute name in `src/main/webapp/WEB-INF/jsp/content/sound/edit.jsp` as well.

jo-elimu commented on src/main/java/ai/elimu/web/content/word/WordCreateController.java:
```diff
@@ -74,13 +74,13 @@ public String handleRequest(Model model, @RequestParam(required = false) String
         model.addAttribute("word", word);
         model.addAttribute("timeStart", System.currentTimeMillis());
-        model.addAttribute("letterSoundCorrespondences", letterSoundDao.readAllOrderedByUsage()); // TODO: sort by letter(s) text
+        model.addAttribute("letterSounds", letterSoundDao.readAllOrderedByUsage()); // TODO: sort by letter(s) text
```
The creation of a new Word at http://localhost:8080/webapp/content/word/create stopped working, since the letter-sound drop-down is empty. Looks like you need to update the frontend file for this controller as well.

<kbd>![](https://github.com/user-attachments/assets/55f51c3f-f78c-40af-a340-b2482d49941b)</kbd>

jo-elimu commented on src/main/java/ai/elimu/web/content/word/WordCreateController.java:
```diff
@@ -110,7 +110,7 @@ public String handleSubmit(
             model.addAttribute("word", word);
             model.addAttribute("timeStart", request.getParameter("timeStart"));
-            model.addAttribute("letterSoundCorrespondences", letterSoundDao.readAllOrderedByUsage()); // TODO: sort by letter(s) text
+            model.addAttribute("letterSounds", letterSoundDao.readAllOrderedByUsage()); // TODO: sort by letter(s) text
```
Same comment here as above.

jo-elimu commented on src/main/java/ai/elimu/web/content/word/WordEditController.java:
```diff
@@ -88,13 +88,13 @@ public String handleRequest(
         model.addAttribute("word", word);
         model.addAttribute("timeStart", System.currentTimeMillis());
-        model.addAttribute("letterSoundCorrespondences", letterSoundDao.readAllOrderedByUsage()); // TODO: sort by letter(s) text
+        model.addAttribute("letterSounds", letterSoundDao.readAllOrderedByUsage()); // TODO: sort by letter(s) text
```
Word editing at http://localhost:8080/webapp/content/word/edit/<ID> stopped working because of this change. So please also update the controller's frontend file at `src/main/webapp/WEB-INF/jsp/content/word/edit.jsp`.

<kbd>![](https://github.com/user-attachments/assets/e9989368-d1a8-4f64-b6a3-864dceed5873)</kbd>

jo-elimu commented on src/main/java/ai/elimu/web/content/word/WordEditController.java:
```diff
@@ -146,7 +146,7 @@ public String handleSubmit(
             model.addAttribute("word", word);
             model.addAttribute("timeStart", request.getParameter("timeStart"));
-            model.addAttribute("letterSoundCorrespondences", letterSoundDao.readAllOrderedByUsage()); // TODO: sort by letter(s) text
+            model.addAttribute("letterSounds", letterSoundDao.readAllOrderedByUsage()); // TODO: sort by letter(s) text
```
Same comment as above.

jo-elimu commented:
@SnehaHS65 I applied some suggestions: https://github.com/elimu-ai/webapp/pull/1725/commits/40cf0e155b6925b14c548688e513dc9bffaa71f9

So please fetch the latest changes before you continue working on the branch.

SnehaHS65 commented:
> @SnehaHS65 I applied some suggestions: [40cf0e1](https://github.com/elimu-ai/webapp/commit/40cf0e155b6925b14c548688e513dc9bffaa71f9)
> 
> So please fetch the latest changes before you continue working on the branch.

@jo-elimu , Can you please check now?

jo-elimu reviewed (changes requested):
@SnehaHS65 See https://github.com/elimu-ai/webapp/pull/1725/files#r1678766292

SnehaHS65 commented:
> @SnehaHS65 See https://github.com/elimu-ai/webapp/pull/1725/files#r1678766292

@jo-elimu , when I fetched from main and merged, it says already up to date that's why I pushed. I don't know why this is happening.

SnehaHS65 commented:
> > @SnehaHS65 See https://github.com/elimu-ai/webapp/pull/1725/files#r1678766292
> 
> @jo-elimu , when I fetched from main and merged, it says already up to date that's why I pushed. I don't know why this is happening.

@jo-elimu , the db script file for version 14 is also changed in main, now what would be the suitable thing to do? should I close this PR?

jo-elimu commented:
> the db script file for version 14 is also changed in main, now what would be the suitable thing to do? should I close this PR?

@SnehaHS65 You can rename the file so that it matches the current snapshot version on the `main` branch: https://github.com/elimu-ai/webapp/blob/main/pom.xml

SnehaHS65 commented:
> > the db script file for version 14 is also changed in main, now what would be the suitable thing to do? should I close this PR?
> 
> @SnehaHS65 You can rename the file so that it matches the current snapshot version on the `main` branch: https://github.com/elimu-ai/webapp/blob/main/pom.xml

@jo-elimu , Yes, I added the new db script. But regarding that log.debug(https://github.com/elimu-ai/webapp/pull/1725/files#r1678766292) I am not sure why it's not appearing in my branch even after pulling from Master. What else do you want me to do in this PR?

SnehaHS65 commented:
@jo-elimu , now all the files should be in sync with the Main branch, can you please review

jo-elimu commented:
> @jo-elimu , now all the files should be in sync with the Main branch, can you please review

Thank you, @SnehaHS65. I will test your code changes later today.

jo-elimu commented on src/main/resources/META-INF/jpa-schema-export.sql:
```diff
@@ -55,6 +55,12 @@
+
     drop table if exists LetterSoundContributionEvent;
 
     drop table if exists LetterSoundCorrespondence;
```
@SnehaHS65 This file looks out-of-date because it shouldn't contain `LetterSoundCorrespondence` anywhere. Try re-generating it by running the `mvn jetty:run` command. And then commit the changes to your branch.

> [!NOTE]
> Note that this file does not get re-generated with `mvn clean compile`. It will only get re-generated after running the webapp on localhost.

SnehaHS65 replied:
Done @jo-elimu

jo-elimu reviewed (changes requested):
Tested webapp on localhost, and this seems to be running well 👍

The `jpa-schema-export.sql` has not been updated to reflect the changes. Needs to be re-generated by running `mvn jetty:run`.

jo-elimu commented on src/main/resources/META-INF/jpa-schema-export.sql:
```diff
@@ -396,18 +385,29 @@
-    create table LetterSoundCorrespondence_Letter (
-       LetterSoundCorrespondence_id bigint not null,
+    create table LetterSound_Letter (
+       LetterSound_id bigint not null,
         letters_id bigint not null,
         letters_ORDER integer not null,
-        primary key (LetterSoundCorrespondence_id, letters_ORDER)
+        primary key (LetterSound_id, letters_ORDER)
     ) engine=MyISAM;
```
@SnehaHS65 FYI, here is the current structure of the `LetterSoundCorrespondence_Letter` table in the production database:

```sql
MariaDB [webapp-HIN]> DESCRIBE LetterSoundCorrespondence_Letter;
+------------------------------+------------+------+-----+---------+-------+
| Field                        | Type       | Null | Key | Default | Extra |
+------------------------------+------------+------+-----+---------+-------+
| LetterSoundCorrespondence_id | bigint(20) | NO   | PRI | NULL    |       |
| letters_id                   | bigint(20) | NO   | MUL | NULL    |       |
| letters_ORDER                | int(11)    | NO   | PRI | NULL    |       |
+------------------------------+------------+------+-----+---------+-------+
```

```sql
MariaDB [webapp-HIN]> SHOW CREATE TABLE LetterSoundCorrespondence_Letter;
+----------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table                            | Create Table

                                                                                                                           |
+----------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| LetterSoundCorrespondence_Letter | CREATE TABLE `LetterSoundCorrespondence_Letter` (
  `LetterSoundCorrespondence_id` bigint(20) NOT NULL,
  `letters_id` bigint(20) NOT NULL,
  `letters_ORDER` int(11) NOT NULL,
  PRIMARY KEY (`LetterSoundCorrespondence_id`,`letters_ORDER`),
  KEY `FK_f3mr6ddn5ctr8edvv9pgtjh2j` (`letters_id`)
) ENGINE=MyISAM DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+----------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

SnehaHS65 replied:
@jo-elimu , Yes added for both LetterSoundCorrespondence_Letter and LetterSoundCorrespondence_Sound in the same db migration script.

jo-elimu reviewed (changes requested):
The DB migration script also needs to include renaming of the `LetterSoundCorrespondence_id` column.

jo-elimu reviewed (approved):
Thank you, @SnehaHS65. I think these changes look good. Please merge, so that we can create a release and try out the migration script on the databases in production (Hindi + Tagalog).

SnehaHS65 commented:
> Thank you, @SnehaHS65. I think these changes look good. Please merge, so that we can create a release and try out the migration script on the databases in production (Hindi + Tagalog).

@jo-elimu , Merged!!
